DB/src/main/java/io/deephaven/db/v2/sources/chunkcolumnsource/CharChunkColumnSource.java

```diff
@@ -0,0 +1,133 @@
+        // if we can slice part of one of our backing chunks, then we will return that instead
+        if (orderedKeys.isContiguous()) {
+            final long firstKey = orderedKeys.firstKey();
+            final int firstChunk = getChunkIndex(firstKey);
+            final int lastChunk = getChunkIndex(orderedKeys.lastKey(), firstChunk);
+            if (firstChunk == lastChunk) {
+                final int offset = (int)(firstKey - offsets.get(firstChunk));
+                final int length = orderedKeys.intSize();
+                final CharChunk<? extends Attributes.Values> charChunk = data.get(firstChunk);
+                if (offset == 0 && length == charChunk.size()) {
+                    return charChunk;
+                }
+                return ((ChunkGetContext)context).resettableCharChunk.resetFromChunk(charChunk, offset, length);
+            }
+        }
+        return getChunkByFilling(context, orderedKeys);
+    }
+
+    @Override
+    public void fillChunk(@NotNull FillContext context, @NotNull WritableChunk<? super Attributes.Values> destination, @NotNull OrderedKeys orderedKeys) {
+        final MutableInt startChunkIndex = new MutableInt(0);
+        final MutableInt destinationOffset = new MutableInt(0);
+        orderedKeys.forAllLongRanges((s, e) -> {
+            while (s <= e) {
+                final int chunkIndex = getChunkIndex(s, startChunkIndex.intValue());
+                final int offsetWithinChunk = (int)(s - offsets.get(chunkIndex));
+                Assert.geqZero(offsetWithinChunk, "offsetWithinChunk");
+                final CharChunk<? extends Attributes.Values> charChunk = data.get(chunkIndex);
+                final int chunkSize = charChunk.size();
+                final long rangeLength = e - s + 1;
+                final int chunkRemaining = chunkSize - offsetWithinChunk;
+                final int length = rangeLength > chunkRemaining ? chunkRemaining : (int) rangeLength;
+                Assert.gtZero(length, "length");
+                destination.copyFromChunk(charChunk, offsetWithinChunk, destinationOffset.intValue(), length);
+                destinationOffset.add(length);
+                startChunkIndex.setValue(chunkIndex + 1);
+                s += length;
+            }
+        });
+    }
+
+    @Override
+    public void fillPrevChunk(@NotNull FillContext context, @NotNull WritableChunk<? super Attributes.Values> destination, @NotNull OrderedKeys orderedKeys) {
+        // immutable, so we can delegate to fill
+        fillChunk(context, destination, orderedKeys);
+    }
+
+    private int getChunkIndex(final long start) {
+        return getChunkIndex(start, 0);
+    }
+
+    private int getChunkIndex(final long start, final int startIndex) {
+        int index = offsets.binarySearch(start, startIndex, offsets.size());
+        if (index < 0) {
+            index = -index - 2;
```

Comment:
This may be mostly about aesthetics, and the more I think about it the more I think you should read it and then do exactly what you want if you don't find it persuasive.

If you store last offset instead of first, you can make this a little prettier:
1. You can use `~1` for the negative index conversion.
2. You can throw an exception if the corrected result is `>= offsets.size()`.

Of course, the trade-off is you'd need to change your calculation around L82:
```
 final int offsetWithinChunk = (int)(s - (chunkIndex == 0 ? 0 :  offsets.get(chunkIndex - 1)));
```
I'm pretty sure this would be a better pattern than relying on the check at L89 to terminate a request for more keys than present, or an AIOOBE from `binarySearch`.

Reply:
I kind of hate the zero check.